File: answers/4/main.py
import numpy as np
import cv2

def otsu_thresh(img):
    max_var = -1
    max_th = 0
    for th in range(1, 254):
        m0 = img[img <= th].mean() if img[img <= th].size != 0 else 0 # mean class 0
        m1 = img[img > th].mean() if img[img > th].size != 0 else 0   # mean class 1
        w0 = img[img <= th].size   # pixel num class 0
        w1 = img[img > th].size    # pixel num class 1
        var = w0 * w1 / ((w0 + w1) ** 2) * ((m0 - m1) ** 2) # inter class variance
        if var > max_var:
            max_th = th
            max_var = var
    _, img = cv2.threshold(img, max_th, 255, cv2.THRESH_BINARY)
    return max_th, img

if __name__ == '__main__':
    img_orig = cv2.imread('questions/dataset/images/imori_256x256.png')
    img_gray = cv2.cvtColor(img_orig, cv2.COLOR_BGR2GRAY)
    th, img_result = otsu_thresh(img_gray)
    print('PYTHON Threshold: {}'.format(th))
    cv2.imshow('python_origin', img_orig)
    cv2.imshow('python_result', img_result)
    cv2.moveWindow('python_result', img_orig.shape[1], 0)
    cv2.waitKey(3000)

File: answers/4/main.cpp
#include <vector>
#include <chrono>
#include <iostream>
#include <opencv2/opencv.hpp>

int otsu_threshold(const cv::Mat &src, cv::Mat &dst) {
    cv::Mat tmp = src.clone();
    if (tmp.channels() == 3) {
        cv::cvtColor(tmp, tmp, cv::COLOR_BGR2GRAY);
    }

    float max_var = -1;
    float max_thr = 0;
    for (int thr = 0; thr < 255; thr++) {
        cv::Mat M0, M1;
        cv::threshold(tmp, M0, thr, 255, cv::THRESH_TOZERO_INV);
        cv::threshold(tmp, M1, thr, 255, cv::THRESH_TOZERO);
        float w0 = cv::sum(cv::countNonZero(M0))[0]; // # of pixels in class 0
        float w1 = cv::sum(cv::countNonZero(M1))[0]; // # of pixels in class 0
        float m0 = cv::sum(cv::sum(M0))[0] / w0; // mean of class 0
        float m1 = cv::sum(cv::sum(M1))[0] / w1; // mean of class 1
        float var = w0 * w1 / pow((w0 + w1), 2) * pow((m0 - m1), 2); // inter class variance
        if (var > max_var) {
            max_thr = thr;
            max_var = var;
        }
    }

    cv::threshold(tmp, dst, max_thr, 255, cv::THRESH_BINARY);
    return max_thr;
}

int main() {
    cv::Mat img_orig = cv::imread("questions/dataset/images/imori_256x256.png");
    cv::Mat img_result;
    int thr = otsu_threshold(img_orig, img_result);
    std::cout << "C++ Threshold: " << thr << std::endl;

    cv::imshow("cpp_origin", img_orig);
    cv::imshow("cpp_result", img_result);
    cv::moveWindow("cpp_result", img_orig.size().width, 0);
    cv::waitKey(3000);
    return 0;
}
